simulator/interface/gdb/gdb_interface.cpp

```diff
@@ -0,0 +1,184 @@
+
+    if (!abfd) {
+        std::cerr << "Input file not set; please re-run GDB with input file" << std::endl;
+        return NULL;
+    }
+
+    int argc = count_argc (argv);
+    SIM_DESC sd = sim_state_alloc (kind, callback);
+
+    try {
+        /* POPL seems to skips first passed argument, so we pass starting at
+         * argv[1], which is "--sysroot=" */
+        config::handleArgs (argc, const_cast<const char **>(&argv[1]));
+
+        /* Create simulator instance */
+        //TODO: add simulator arguments
+        sd->instanceId = simInstances.size ();
+        simInstances.emplace_back (Simulator::create_simulator ("mips32", true, true),
+                                   abfd->filename, sd->instanceId);
+    }
+    catch (const config::HelpOption &e) {
+        std::cout << "Functional simulator for MIPS-based CPU (GDB)"
+                  << std::endl << std::endl << e.what () << std::endl;
+        sim_state_free (sd);
+        return NULL;
+    }
+    catch (const std::exception &e) {
+        std::cerr << e.what () << std::endl;
+        sim_state_free (sd);
+        return NULL;
+    }
+    catch (...) {
+        std::cerr << "Unknown exception\n";
+        sim_state_free (sd);
+        return NULL;
+    }
+
+    std::cout << "MIPT-MIPS simulator instance created, id " << sd->instanceId << std::endl;
+
+    return sd;
+}
+
+
+void sim_close (SIM_DESC sd, int) {
+    simInstances.at (sd->instanceId).ptr.reset ();
+    sim_state_free (sd);
+}
+
+
+SIM_RC sim_load (SIM_DESC sd, const char *, struct bfd *, int) {
+    SimulatorInstance &simInst = simInstances.at (sd->instanceId);
+    simInst.ptr->gdb_load (simInst.filename);
+    return SIM_RC_OK;
+}
+
+
+SIM_RC sim_create_inferior (SIM_DESC sd, struct bfd *,
+                            char *const *, char *const *) {
+    simInstances.at (sd->instanceId).ptr->gdb_prepare ();
+    return SIM_RC_OK;
+}
+
+
+int sim_read (SIM_DESC sd, SIM_ADDR mem, unsigned char *buf, int length) {
+    SimulatorInstance &simInst = simInstances.at (sd->instanceId);
+    return static_cast<int> (simInst.ptr->gdb_mem_read (mem, buf, static_cast<size_t> (length)));
+}
+
+
+int sim_write (SIM_DESC sd, SIM_ADDR mem, const unsigned char *buf, int length) {
+    SimulatorInstance &simInst = simInstances.at (sd->instanceId);
+    return static_cast<int> (simInst.ptr->gdb_mem_write (mem, buf, static_cast<size_t> (length)));
+}
+
+
+int sim_fetch_register (SIM_DESC sd, int regno, unsigned char *buf, int length) {
+    (void) sd;
+    (void) regno;
+    (void) buf;
+    (void) length;
+    return 0;
+}
+
+
+int sim_store_register (SIM_DESC sd, int regno, unsigned char *buf, int length) {
+    (void) sd;
+    (void) regno;
+    (void) buf;
+    (void) length;
+    return 0;
+}
+
+
+void sim_info (SIM_DESC sd, int verbose) {
+    (void) sd;
+    (void) verbose;
+}
+
+
+void sim_resume (SIM_DESC sd, int step, int) try {
+    simInstances.at (sd->instanceId).ptr->gdb_resume (step);
+}
+catch (BearingLost) {
```

Comment:
It feels you have small experience with exceptions, so let me show usual way to catch and handle them:
```c++
catch (const BearingLost &e) {
// Always catch by reference.
// Actual exception may be derived from BearingLost,
// and if it is catched by value, it would be truncated to the base class
  std::cerr // better to use it for errors
   <<  e.what() //  That would print the message of exception
   << std::endl;
}